pkg/iscsi/iscsi_util.go

```
@@ -109,8 +109,9 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, targetPath string) error
 		return err
 	}
 
-	iscsi_lib.Disconnect(connector.TargetIqn, connector.TargetPortals)
-
+	if disConnectErr := iscsi_lib.Disconnect(connector.TargetIqn, connector.TargetPortals); disConnectErr != nil {
+		klog.Warningf("Warning: Disconnect failed for IQN: %v", connector.TargetIqn)
```

Is there a scenario where disconnect fails but we want to continue? Would this potentially leak connections?

---

Yes, we should always continue when dealing with iscsi disconnects IMO, and in fact we are; this change only add a log warning if the disconnect fails but it doesn't change or stop the disconnect sequence.

---

Yep, we are just capturing and moving on.

---

What I'm wondering about is if we should be retrying to disconnect after the first failed attempt. Could we end up leaking a bunch of connections after some time and cause other issues down the line?

---

@humblec is this resolved? Should we be returning an error and failing instead of continuing? This is the error sequence I'm concerned about:

1. Disconnect fails. We log and move on
2. Next step to `os.Remove` succeeds.
3. Since NodeUnpublishVolume succeeds we never come back and try to clean up this connection and the LUN still remains attached to this node.

The reason why I'm concerned about this is because we encountered data corruption in in-tree fc plugin due to the volume not being successfully cleaned up from the node and a pod rescheduled to another node and then rescheduled again back to the original node. However the disk was not reattached since it was already connected and the pod ended up updating stale data causing corruption. https://github.com/kubernetes/kubernetes/issues/101791

A couple more questions about this disconnect and general unpublish flow:
- Can you have multiple volumes on the same node sharing the same connection? Would disconnecting here cause other mounted volumes to fail?
- Does disconnect also detach volumes from the node?

cc @jsafrane

---

@msau42 there are some good amount of changes required in the csi-lib-iscsi which has been done at my local test , but not part of this PR. There is also a PR (https://github.com/kubernetes-csi/csi-lib-iscsi/pull/29)  from @paullaffitte  which carry almost the same fixes .  I have a discussion  scheduled with Paul about coordinating the effort and to get the library in a good shape first so that it can be pulled over here. 

Cc @xing-yang @jsafrane @saad-ali 

I will revert on the final action items to take these forward. 

Till then we can hold this PR.